Add synchronous `wxWebRequest`

vadz opened the pull request:
This is ready for merging as is, but I'm probably going to fix a few other minor things in this branch yet. Still, any comments about the API and testing would be already very appreciated.

In addition to the main topic, there are a couple of other changes here:

- Allow more flexibility when disabling TLS checks.

TODO:

- [x] WinHTTP backend
- [x] `NSURLSession` backend (although the current implementation is suboptimal).
- [x] Documentation

vadz committed:
Default special member functions of wxWebRequest etc

No real changes, just replace the trivial implementations of default and
copy ctors, assignment operators and dtors of wxWebRequest and related
classes with the compiler-generated versions, which is shorter, more
clear and marginally more efficient.

vadz committed:
Factor out wxWebSessionBase from wxWebSession

No real changes yet, just factor out the part of wxWebSession that will
be reused by wxWebSessionSync.

vadz committed:
Initialize members of wxWebRequest etc in their declarations

No real changes, just simplify the initialization code and ensure that
members are always initialized correctly, even if/when new ctors are
added.

vadz committed:
Inline wxWebRequestCURL::DestroyHeaderList()

This trivial function was called only once and it doesn't seem to be
worth having it, especially because performing cleanup in the dtor makes
the code easier to follow.

This also allows to remove the line resetting m_headerList to null
pointer, as it's now clear that it's not needed.

No real changes.

vadz committed:
Add comment explaining WX_TEST_WEBREQUEST_URL_SELF_SIGNED

Give an example URL it can be set to for testing.

vadz committed:
Implement sync requests support in libcurl backend

Refactor the existing code to allow adding wxWebSessionSync which uses
an "easy" CURL handle instead of CURLM used by async wxWebSession.

Mostly replace functions indirectly changing state by calling SetState()
with the functions explicitly returning a Result and then use
CheckResult() to update the state if necessary in the async version of
the code.

vadz committed:
Avoid gcc 4.8 warnings for unused parameters in default functions

Unlike all the other compilers and versions of gcc, this one warns about
unused parameters in the compiler-generated functions, so just remove
their names to avoid the warning.

vadz committed:
Call WriteData() unconditionally on send request completion

No real changes, just simplify the code a bit by always calling
WriteData() instead of checking whether we have any data to write: this
makes this case more consistent with the "write request completion" one
as we didn't check if there was any data remaining there, and we can
also avoid doing it here because WriteData() does it internally anyhow.

vadz committed:
Replace wxWebResponseImpl::Init() with InitFileStorage()

The old function name wasn't very descriptive and even misleading, as
Init() is typically used for simple initializers in wxWidgets, but this
function was anything but, because it could call SetState() in case of
error.

Remove this side effect and just return a Result from it and handle it
in the caller.

This also probably fixes a bug in NSURLSession backend, as it simply
ignored the failure in Init() before and continued with the request
normally instead of abandoning it.

vadz committed:
Remove wxWebResponseWinHTTP::ReportAvailableData()

This trivial function was called only once and just complicated the code
unnecessarily, so get rid of it by making ReportDataReceived() public
and calling it directly.

No real changes.

vadz committed:
Test reading more bytes than one buffer size

Verify that we are not limited to the buffer size when receiving the
remote document.

vadz committed:
Implement sync requests support in WinHTTP backend

vadz committed:
Partially implement sync requests support in NSURLSession backend

This version uses completion handler-based approach for implementing
synchronous wxWebRequest in wxOSX which is relatively simple, but
unfortunately doesn't allow specifying the authentication information,
so this is probably not the right approach and it will have to be
rewritten.

vadz committed:
Factor out wxWebRequestBase from wxWebRequest too

Similar to the previous commit, extract non-async-related functionality
of wxWebRequest into a new base class in order to allow adding
wxWebRequestSync deriving from it in the upcoming commits.

No real changes yet.

vadz committed:
Add wxWebRequest::Result class

This class combines wxWebRequest::State and the error message and will
be used as return value for the synchronous functions that will be added
soon, while allowing to reuse them and their return value in
asynchronous case as well thanks to the CheckResult() and HandleResult()
helpers.

vadz committed:
Add wxWebSessionSync and wxWebRequestSync stabs

This commit just defines the API and adds the documentation.

Note that we avoid documenting wxWebRequestBase and wxWebRequestBase
classes that are implementation details, even if it means that we
duplicate the descriptions of most of the functions between the async
and sync versions of the classes -- but this is actually useful as it
allows to document things working differently in the sync case.

vadz committed:
Add unit tests for wxWebRequestSync

They mirror the existing async tests.

vadz committed:
Factor out GetResultFromHTTPStatus() helper

This was previously part of GetResultFromHTTPStatus(), but make it
available as a standalone function now.

Also don't bother with handling null response in this function, as it
could be never called with a null pointer.

vadz committed:
Add support to wxWebRequestSync to the common code

This temporarily breaks the build due to the addition of new pure
virtual functions, but this will be temporarily fixed by the next commit
and then really fixed later.

vadz committed:
Provide temporary stubs for the new pure virtual functions

This is done just to avoid breaking the build until the new API is
implemented and this commit will be reverted later.

vadz committed:
Revert "Provide temporary stubs for the new pure virtual functions"

This reverts commit 3502adcbb3709a849ae6da75b6c2d7858599b05b as the new
API is now implemented in all backends.

vadz committed:
Add WebRequest::Sync::Manual test

This allows to run wxWebRequestSync::Execute() manually too.

This commit is best viewed with Git --color-moved option.

vadz committed:
Stop using wxWebAuthChallenge with wxWebRequestSync

The wxWebAuthChallenge-based approach seems to be unnecessarily
complicated for the synchronous use case and isn't supported by the
current NSURLSession backend implementation (although this could be
changed by rewriting it to use a different delegate for the sync
requests and remembering wxWebAuthChallenge in it between the calls to
Execute()).

For now, just drop support for it and require specifying the
authentication information in the URL itself.

vadz committed:
Implement support for credentials in the URL in WinHTTP too

This was already supported in the other backends, but not in this one,
resulting in inconsistent behaviour between platforms.

Fix this by adding support for passing username:password directly in the
URL -- which will be especially useful for synchronous requests.

vadz committed:
Test response body in authentication tests too

Check that we do receieve the expected body and not just the correct
status.

vadz committed:
Add tests for specifying credentials in URL for async requests

This didn't work in WinHTTP backend before but now that it does, add
tests checking this.

vadz committed:
Support credentials in the URL in WinHTTP wxWebRequestSync too

This required slightly refactoring SetCredentials() to allow using it
from synchronous code.

This commit is best viewed ignoring whitespace-only changes.

vadz commented:
I have more changes planned (while working on this, I've found several other problems...), but I think this might be in a reasonably good shape by now.

As proposed in my [post to wx-dev](https://groups.google.com/g/wx-dev/c/WHIMLdQMzAE/m/zkq0wTn9BgAJ) I've removed support for using `wxWebAuthChallenge` in the synchronous case, as it just seems gratuitously complicated when we can simply put the username and password in the URL. The main drawback of this is that we can't support any other auth schemes, not using passwords, but then we didn't support them before anyhow, so it's not really a loss.

Also, this uncovered the fact that using username and password worked for (async) `wxWebRequest` in libcurl and NSURLSession but _not_ WinHTTP backend, which clearly seems wrong, so I've fixed this too and if you know the username/password in advance you can just put them in the URL instead of bothering with providing them later.

vadz commented:
Huh, I spoke too soon, `WebRequest::Sync::Auth::Digest` test fails under Ubuntu 18.04 even though it succeeds under later versions. This is pretty weird, especially as the same version of go-httpbin is used everywhere. Must be due to some difference in system libraries, maybe libcurl under 18.04 is too old and doesn't support credentials for the digest authentication in the URL?

vadz committed:
Fix tests for response body after authentication

It turns out different httpbin clones produce slightly different
responses, so accept any of them.

vadz committed:
Skip digest auth test with credentials in URL under Ubuntu 18.04

Instead of dealing with this obsolete system, just skip the test there.

Blake-Madden commented:
`wxWebRequestSync::Execute` states "This function blocks for potentially long time and so must not be used from the main thread."

To be truly synchronous then, one would need to launch this in a `std::jthread` and join it. But then that would defeat the point if `wxWebRequestSync::Execute` gets stuck somehow.

Perhaps a timeout parameter (in milliseconds) could be added to `wxWebRequestSync::Execute`?

vadz commented:
There is no way to specify a timeout with blocking API (such as `curl_easy_perform()` or various `WinHttpXXX()` functions in synchronous mode), so I don't see how can this be done. We'd need a third mode, using async API in a pseudo-blocking way, without dispatching the events, in order to implement this.

vadz committed:
Refactor manual URL tests to avoid code duplication

Add InitManualRequest() to the fixture to avoid having the same code for
the async and sync cases.

No real changes, this is a pure refactoring.

vadz committed:
Show error message in the manual URL tests

This can provide more information about the failure reasons.

vadz committed:
Document wxWebRequestSync::DisablePeerVerify() macOS limitation

It doesn't work currently with synchronous requests due to not using a
delegate for them.

vadz committed:
Add more flexible way of disabling SSL checks

Allow disabling host name checks too, to make the connection completely
insecure by accepting any certificate for any host.

Incidentally, this also allows checking for valid certificate for any
host, even if this doesn't seem tremendously useful.

Document that disabling any kind of verification in NSURLSession-based
backend disables all of them and also significantly refactor and extend
unit tests checking for this.

vadz committed:
Simplify passing URLs for SSL testing to unit test

Allow defining a single WX_TEST_WEBREQUEST_USE_BADSSL environment
variable to use badssl.com for all tests instead of bothering with
defining each variable separately (but it's still possible to do it and
override the default values implied by WX_TEST_WEBREQUEST_USE_BADSSL).

vadz commented:
Just in case this went unnoticed, I've updated the description after pushing the latest changes.

vadz committed:
Remove unused wxWebResponseImpl::m_readSize

This variable was never changed from its initial value and was only used
in WinHTTP backend, so remove it and replace its occurrences in WinHTTP
code with the constant that was previously used to initialize it.

No real changes.

vadz committed:
Use the same logic for HTTP methods in all backends

No real changes, just add GetHTTPMethod() helper function in the base
wxWebRequestImpl class and use it in all backends: previously, WinHTTP
and NSURLSession ones already did the same thing, but each in their own
way, while libcurl backend used a different logic which was mostly
equivalent, but this was not really obvious, while now it should be.

Also add a test for using a custom method (DELETE), for now without
body, as this is (still) not supported by libcurl backend.

vadz committed:
Add WebRequest::Version pseudo-test to see the backend version

This can be notably useful to see libcurl version.

vadz committed:
Fix libcurl backend to handle body with custom requests correctly

Previously body wasn't sent, even if it was specified, for the custom
requests such as DELETE.

Fix this by setting CURLOPT_UPLOAD and CURLOPT_INFILESIZE options if the
body is present for any method except POST, which is handled specially
by libcurl, and not only for PUT.

Add a unit test checking that this works now.

vadz committed:
Return upper case string from GetHTTPMethod()

This doesn't matter for WinHTTP and URLSession backends, but simplifies
code in the libcurl one.

No real changes.

vadz committed:
Remove unnecessary IsAsync() checks from SetCredentials()

This function is not called at all for synchronous requests any longer,
so don't bother checking the request is async in it.

No real changes.

vadz commented:
I plan to merge this soon and open a separate PR with proxy-related changes.

vadz committed:
Make setting options in libcurl backend safer

No real changes, just make it impossible to mistakenly pass a wxString
or wxCharBuffer to unsafe variadic curl_easy_setopt() and use our own
safe wxCURLSetOpt() instead.

This also simplifies the code a bit.

vadz committed:
Don't crash in EnablePersistentStorage() if session is invalid

Add a check for the session validity to this function for consistency
with all the other member functions of wxWebSession.

vadz committed:
Add logging for auth requests in WinHTTP backend

Give more information in the logs to make them easier to follow.

vadz committed:
Improve logging when reading data in WinHTTP backend

Don't log "receiving 18446744073709551615 bytes" when we have no content
length at all, this wasn't very readable.

vadz committed:
Verify that requests are created successfully in the tests

This may not happen if the session couldn't be initialized for some
reason and it's better to stop the test immediately in this case than to
give a lot of other confusing errors later.

vadz committed:
Improve wxWebSessionURLSession::EnablePersistentStorage() check

Don't complain if EnablePersistentStorage(false) is called, as the error
message doesn't make much sense in this case.

No real changes.
